File: ServerAspNet/Lobby/CoopLobby.cs
using System.Collections.Concurrent;
using SuperMarioOdysseyOnline.Server.Packets;

namespace SuperMarioOdysseyOnline.Server.Lobby;

[Lobby("Coop")]
internal class CoopLobby(Guid id, string name) : ILobby
{
    private readonly ConcurrentDictionary<Guid, IPlayer> _players = new();

    public Guid Id => id;

    public string Name => name;

    public CoopLobby(LobbyDetails lobbyDetails)
        : this(Guid.NewGuid(), lobbyDetails.Name)
    {
    }

    public IEnumerable<IPlayer> Players => _players.Select(pair => pair.Value);

    public IPlayer GetOrAddPlayer(Guid id) => _players.GetOrAdd(id, (x) => new Player(x));

    public void HandleReceivedPacket(IPacket packet)
    {
    }
}

File: SuperMarioOdysseyOnline.Server.Application/wwwroot/js/CoopLobby.js
import LobbyType from "./LobbyType.js";

/**
 * @typedef {import('./Types').ILobby} ILobby
 */

/**
 * @implements {ILobby}
 */
class CoopLobby {
  /** @type {string} */
  #id;
  get id() {
    return this.#id;
  }

  /** @type {string} */
  #name;
  get name() {
    return this.#name;
  }

  get type() {
    return LobbyType.Coop;
  }

  constructor(
    /** @type {string} */
    id,
    /** @type {string} */
    name
  ) {
    this.#id  = id;
    this.#name  = name;
  }
}

export default CoopLobby;
